src/util/assignment.rs

```diff
@@ -0,0 +1,263 @@
+use core::str;
+use std::{collections::{BTreeMap, HashMap}, io::Read};
+
+use crypto_box::{
+    aead::{Aead, AeadCore, OsRng},
+    SalsaBox, PublicKey, SecretKey
+};
+use serde::{Deserialize, Serialize};
+use serde_json::Value;
+use serde_with::serde_as;
+use serde_with::base64::Base64;
+use sha2::Sha512;
+use sha2::Digest;
+use sha3::digest::generic_array::GenericArray;
+use curve25519_dalek::edwards::CompressedEdwardsY;
+use flate2::read::GzDecoder;
+
+use crate::util::signature::timed_hmac_now;
+
+#[derive(Debug, Serialize, Deserialize, Clone)]
+#[serde(rename_all = "camelCase")]
+pub struct Chunk {
+    pub id: String,
+    pub base_url: String,
+    pub files: HashMap<String, String>,
+    pub size_bytes: u64,
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub struct Dataset {
+    pub id: String,
+    pub base_url: String,
+    pub chunks: Vec<Chunk>,
+}
+
+#[derive(Debug, Default, Serialize, Deserialize)]
+#[serde(rename_all = "kebab-case")]
+struct Headers {
+    worker_id: String,
+    worker_signature: String,
+}
+
+#[serde_as]
+#[derive(Debug, Default, Serialize, Deserialize, Clone)]
+#[serde(rename_all = "camelCase")]
+struct EncryptedHeaders {
+    #[serde_as(as = "Base64")]
+    identity: Vec<u8>,
+    #[serde_as(as = "Base64")]
+    nonce: Vec<u8>,
+    #[serde_as(as = "Base64")]
+    ciphertext: Vec<u8>,
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+#[serde(rename_all = "camelCase")]
+struct WorkerAssignment {
+    status: String,
+    chunks_deltas: Vec<u64>,
+    encrypted_headers: EncryptedHeaders,
+}
+
+#[derive(Debug, Default, Serialize, Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub struct Assignment {
+    datasets: Vec<Dataset>,
+    worker_assignments: HashMap<String, WorkerAssignment>,
+    #[serde(skip)]
+    chunk_map: Option<HashMap<String, u64>>,
+    #[serde(skip)]
+    pub id: String,
+}
+
+#[derive(Serialize, Deserialize)]
+pub struct NetworkAssignment {
+    pub(crate) url: String,
+    pub(crate) id: String,
+}
+
+#[derive(Serialize, Deserialize)]
+pub struct NetworkState {
+    pub(crate) network: String,
+    pub(crate) assignment: NetworkAssignment
+}
+
+impl Assignment {
+    pub fn add_chunk(&mut self, chunk: Chunk, dataset_id: String, dataset_url: String) {
+        match self.datasets.iter_mut().find(|dataset| dataset.id == dataset_id) {
+            Some(dataset) => dataset.chunks.push(chunk),
+            None => self.datasets.push(Dataset { id: dataset_id, base_url: dataset_url, chunks: vec![chunk] }),
+        }
+        self.chunk_map = None
+    }
+
+    pub async fn from_url(url: String, previous_id: Option<String>) -> Result<Option<Self>, Box<dyn std::error::Error>> {
```

Comment:
Let's use `anyhow::Error` instead of `Box<dyn std::Error>`, it's much more convenient for error conversions

Reply:
thanks, didn't know anyhow::Error may work this way